lib/datasource/artifactory/index.ts

```diff
@@ -0,0 +1,97 @@
+import { logger } from '../../logger';
+import { ExternalHostError } from '../../types/errors/external-host-error';
+import { cache } from '../../util/cache/package/decorator';
+import { parse } from '../../util/html';
+import { HttpError } from '../../util/http/types';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { datasource, defaultRegistryUrl } from './common';
+
+export class ArtifactoryDatasource extends Datasource {
+  static readonly id = datasource;
+
+  constructor() {
+    super(datasource);
+  }
+
+  override readonly customRegistrySupport = true;
+
+  override readonly defaultRegistryUrls = [defaultRegistryUrl];
+
+  override readonly caching = true;
+
+  override readonly registryStrategy = 'merge';
+
+  @cache({
+    namespace: `datasource-${datasource}`,
+    key: ({ registryUrl }: GetReleasesConfig) => `${registryUrl}`,
+  })
+  async getReleases({
+    lookupName,
+    registryUrl,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    if (registryUrl === defaultRegistryUrl) {
+      logger.warn(
+        'artifactory datasource requires custom registryUrl. Skipping datasource'
+      );
+      return null;
+    }
+
+    const url = `${registryUrl}/${lookupName}`;
+
+    const result: ReleaseResult = {
+      homepage: 'https://jfrog.com/artifactory',
+      releases: [],
+    };
+    try {
+      const response = await this.http.get(url);
+      const body = parse(ArtifactoryDatasource.cleanSimpleHtml(response.body));
+      const nodes = body.querySelectorAll('a');
+
+      let candidates: string[] = [];
+      nodes.forEach((node) => candidates.push(node.innerHTML));
+
+      // filter out hyperlink to navigate to parent folder
+      candidates = candidates.filter(
+        (candidate) => candidate !== '../' && candidate !== '..'
+      );
+
+      candidates.forEach((candidate) => {
+        const parsedCandidate: string =
+          candidate.slice(-1) === '/' ? candidate.slice(0, -1) : candidate;
+
+        const thisRelease: Release = {
+          version: '',
+        };
+        thisRelease.version = parsedCandidate;
+        result.releases.push(thisRelease);
+      });
+
+      if (result.releases.length) {
+        logger.debug('Found ' + String(result.releases.length) + ' releases');
+      } else {
+        logger.debug(
+          'Not found any version of ' + lookupName + ' under ' + url
+        );
+      }
+    } catch (err) {
+      // istanbul ignore else: not testable with nock
+      if (err instanceof HttpError) {
+        if (err.response?.statusCode !== 404) {
+          throw new ExternalHostError(err);
+        }
+      }
+      this.handleGenericErrors(err);
+    }
+
+    return result.releases.length ? result : null;
+  }
+
+  private static cleanSimpleHtml(html: string): string {
+    return (
+      html
+        // preformatted text hides the "a" nodes otherwise
```

Comment:
Is this a limitation of the html parser we're using?

Reply:
Not sure, it could be. When I first forked renovate, the response was parsed out of the box. I found that the html parser changed in https://github.com/renovatebot/renovate/pull/8909. The html files under _\_\_fixture\_\__ folder shows an example of the response.

Reply:
That pr was merge long ago. Does the current parser makes problems? Then we should open an issue on upstream repo to get that fixed, instead of adding ugly workarounds.

You should always make sure your fork is uptodate before any development, otherwise you risk to redo all you work again and again, with will be very frustrating for all

Reply:
Parsing the html files from the _\_\_fixture\_\__ folder in online html viewers works fine and the children nodes are detected. For example: https://jsonformatter.org/html-viewer
Thus, @viceice may be right and there is a bug in the [node-html-parser](https://github.com/taoqf/node-html-parser). I have tried the latest version (4.1.4) and does not work either. I'll open an issue as suggested.

Wrt. the update, I forked this repository like a year ago and by that time the _parser_ and _querySelectorAll('a')_ was returning all the children nodes. I just wanted to point that out to bring some insight to this topic. A week ago, when I started contributing to the official repository, I updated my fork and again yesterday, right before creating this pr as recommended.

Reply:
Found an old issue in node-html-parser: https://github.com/taoqf/node-html-parser/issues/78
There it is explained how to add some configuration to process nodes around preformatted text.
I have applied the recommendation in the last commit.